engine-comparison/report-gen/generate-report.go

```diff
@@ -0,0 +1,175 @@
+		if fd.IsDir() {
+			out_ls = append(out_ls, fd)
+		}
+	}
+	return
+}
+
+// Extend "records" matrix to have rows until time "desired_time"
+// Return: Extended version of record
+func extendRecordsToTime(records [][]string, desired_time int, record_cols int) [][]string {
+	lenr := len(records)
+	// records[1] stores cycle [1], as records[0] is column names
+	for j := lenr; j < desired_time+1; j++ {
+		records = append(records, make([]string, record_cols))
+		records[j][0] = strconv.Itoa(j)
+	}
+	return records
+}
+
+// Handles the CSV Reader for a single trial, and updates records[][] accordingly. Returns the updated records
+func handleTrialCSV(this_reader *csv.Reader, records [][]string, column_name string, num_record_columns int, j int) [][]string {
+	// Read whole CSV to an array
+	experiment_records, err := this_reader.ReadAll()
+	checkErr(err)
+	// Add the name of this new column to records[0]
+	records[0] = append(records[0], column_name)
+
+	final_time, err := strconv.Atoi(experiment_records[len(experiment_records)-1][0])
+	checkErr(err)
+	//If this test went longer than all of the others, so far
+	if len(records) < final_time+1 {
+		records = extendRecordsToTime(records, final_time, num_record_columns)
+	}
+	for _, row := range experiment_records {
+		// row[0] is time, on the x-axis; row[1] is value, on the y-axis
+		time_now, err := strconv.Atoi(row[0])
+		checkErr(err)
+		records[time_now][j+1] = row[1]
+	}
+	return records
+}
+
+func handleFengine(fengine os.FileInfo, current_path string, desired_report_fname string) [][]string {
+	// Create matrix, to eventually become a CSV
+	records := [][]string{{"time"}}
+
+	// Enter sub-directories
+	potential_trials, err := ioutil.ReadDir(path.Join(current_path, fengine.Name()))
+	checkErr(err)
+	trials := onlyDirectories(potential_trials)
+
+	num_record_columns := len(trials) + 1
+	for j, trial := range trials {
+		// Create fds
+		this_file, err := os.Open(path.Join(current_path, fengine.Name(), trial.Name(), desired_report_fname))
+		checkErr(err)
+		this_reader := csv.NewReader(this_file)
+
+		records = handleTrialCSV(this_reader, records, fengine.Name()+"-"+trial.Name(), num_record_columns, j)
+		this_file.Close()
+	}
+
+	this_fe_file, err := os.Create(path.Join(current_path, fengine.Name(), desired_report_fname))
+	checkErr(err)
+	this_fe_writer := csv.NewWriter(this_fe_file)
+	this_fe_writer.WriteAll(records)
+	this_fe_file.Close()
+	return records
+}
+
+// Identify the fastest trial  column in records and add it to meta_records
+func appendFastestTrial(meta_records [][]string, records [][]string) [][]string {
+	// Initialized to false: track whether this trial has lasted a long time
+	trials := make([]bool, len(records[0])-1)
+	// Use int for faster comparisons
+	finished_trials := len(trials)
+	// Speed of Fastest Trial
+	var speed_of_ft int
+	var fastest_trial int
+
+	// Find the fastest trial by working backwards, since data points are sometimes dropped
```

Comment:
When are data points dropped?  What if data points at the end are dropped?

Reply:
Data is dropped in the runner if `gsutil` takes more time than one cycle to sync the corpus. Data points at the end shouldn't be dropped, because that's the last `rsync` before resetting for another trial. However, looking at `runner.sh` now, I think this last cycle might accidentally be dropped every time

Reply:
Is the issue that the second `gsutil rsync` overrides the first or that the dispatcher doesn't check the same cycle again if `gsutil` is still going?

Reply:
Sort of the first; `gsutil` doesn't run in the background on runners, so the while loop doesn't continue to snapshot the next corpus until `gsutil` finishes. If it did run in the background, there would be the possibility of two `gsutil rsync`s running at the same time, trying to sync two different files to the same directory. Which could perhaps work, but would likely just further overload internet bandwidth

Reply:
Ok, I see.